mailpoet/lib/Automation/Integrations/MailPoet/Triggers/OrderStatusChangedTrigger.php

```diff
@@ -0,0 +1,138 @@
+use MailPoet\Subscribers\SubscribersRepository;
+use MailPoet\Validator\Builder;
+use MailPoet\Validator\Schema\ObjectSchema;
+use MailPoet\WooCommerce;
+use MailPoet\WP\Functions;
+
+class OrderStatusChangedTrigger implements Trigger {
+
+
+  /** @var Functions */
+  private $wp;
+
+  /** @var SubscribersRepository */
+  private $subscribersRepository;
+
+  /** @var Segments\WooCommerce */
+  private $woocommerce;
+
+  /** @var WooCommerce\Helper */
+  private $woocommerceHelper;
+
+  public function __construct(
+    Functions $wp,
+    SubscribersRepository $subscribersRepository,
+    Segments\WooCommerce $wooCommerce,
+    WooCommerce\Helper $woocommerceHelper
+  ) {
+    $this->wp = $wp;
+    $this->subscribersRepository = $subscribersRepository;
+    $this->woocommerce = $wooCommerce;
+    $this->woocommerceHelper = $woocommerceHelper;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:order-status-changed';
+  }
+
+  public function getName(): string {
+    return __('Order status changed', 'mailpoet');
+  }
+
+  public function getArgsSchema(): ObjectSchema {
+
+    return Builder::object([
+      'status' => Builder::string()->required()->default('wc-completed'),
+    ]);
+  }
+
+  public function getSubjectKeys(): array {
+    return [
+      SubscriberSubject::KEY,
+      OrderSubject::KEY,
+    ];
+  }
+
+  public function validate(StepValidationArgs $args): void {
+    // TODO: Implement validate() method.
+  }
+
+  public function registerHooks(): void {
+    add_action(
+      'woocommerce_order_status_changed',
+      [
+        $this,
+        'handle',
+      ],
+      10,
+      3
+    );
+  }
+
+  public function handle(int $orderId, string $oldStatus, string $newStatus): void {
+    $order = $this->woocommerceHelper->wcGetOrder($orderId);
+    if (!$order instanceof \WC_Order) {
+      return;
+    }
+
+    $subscriber = $this->findOrCreateSubscriber($order);
+    if (!$subscriber instanceof SubscriberEntity) {
+      return;
+    }
+
+    $this->wp->doAction(Hooks::TRIGGER, $this, [
+      new Subject(OrderSubject::KEY, ['order_id' => $order->get_id()]),
+      new Subject(SubscriberSubject::KEY, ['subscriber_id' => $subscriber->getId()]),
+    ]);
+  }
+
+  private function findOrCreateSubscriber(\WC_Order $order): ?SubscriberEntity {
+
+    $subscriber = $this->findSubscriber($order);
+    if ($subscriber) {
+      return $subscriber;
+    }
+
+    $this->woocommerce->synchronizeGuestCustomer($order->get_id());
+
+    return $this->findSubscriber($order);
+  }
+
+  private function findSubscriber(\WC_Order $order): ?SubscriberEntity {
+    $subscriber = $this->subscribersRepository->findOneBy(['email' => $order->get_billing_email()]);
+    if ($subscriber instanceof SubscriberEntity) {
+      return $subscriber;
+    }
+
+    $subscriber = $this->subscribersRepository->findOneBy(['wpUserId' => $order->get_user_id()]);
+    if ($subscriber instanceof SubscriberEntity) {
+      return $subscriber;
+    }
```

Comment:
It's probably a small edge case, but shouldn't the `wpUserId` go first, so we try to match the most accurately possible? That said, it could then happen that an email would be sent to a different email address than the one that was used during the order, right? 🤔

Reply:
> email would be sent to a different email address than the one that was used during the order, right? 
Yes. They can differ.